src/lib/notification.vala

```diff
@@ -0,0 +1,254 @@
+
+			// Set the priority
+			if ((variant = hints.lookup("urgency")) != null && variant.is_of_type(VariantType.BYTE)) {
+				this.urgency = (NotificationUrgency) variant.get_byte();
+			}
+
+			// Set the category
+			if ((variant = hints.lookup("category")) != null) {
+				this.category = variant.get_string();
+			}
+
+			// Set the application ID and app info
+			if ((variant = hints.lookup("desktop-entry")) != null && variant.is_of_type(VariantType.STRING)) {
+				this.app_id = variant.get_string();
+				app_id.replace(".desktop", "");
+				this.app_info = new DesktopAppInfo("%s.desktop".printf(app_id));
+			}
+
+			bool image_found = false;
+
+			// Per the Freedesktop Notification spec, first check if there is image data
+			if (
+				(variant = hints.lookup("image-data")) != null ||
+				(variant = hints.lookup("image_data")) != null
+			) {
+				var pixbuf = decode_image(variant);
+				if (pixbuf != null) {
+					image = new Gtk.Image.from_pixbuf(pixbuf);
+					image_found = true;
+				}
+			}
+
+			// If there was no image data, check if we have a path to the image to use.
+			if (!image_found) {
+				if (
+					(variant = hints.lookup("image-path")) != null ||
+					(variant = hints.lookup("image_path")) != null
+				) {
+					var path = variant.get_string();
+
+					if (Gtk.IconTheme.get_default().has_icon(path) && path != this.app_icon) {
+						var icon = new ThemedIcon(path);
+						this.image = new Gtk.Image.from_gicon(icon, Gtk.IconSize.DIALOG);
+						image_found = true;
+					} else if (path.has_prefix("/") || path.has_prefix("file://")) {
+						try {
+							var pixbuf = new Gdk.Pixbuf.from_file(path);
+							this.image = new Gtk.Image.from_pixbuf(pixbuf);
+							image_found = true;
+						} catch (Error e) {
+							critical("Unable to get pixbuf from path: %s", e.message);
+						}
+					}
+				}
+			}
+
+			// If no image path, try the app_icon parameter.
+			if (!image_found) {
+				if (app_icon != "" && !app_icon.contains("/")) { // Use the app icon directly
+					this.image = new Gtk.Image.from_icon_name(app_icon, Gtk.IconSize.DIALOG);
+					image_found = true;
+				} else if (app_icon == "" && app_info != null) { // Try to get icon from application info
+					this.image = new Gtk.Image.from_gicon(app_info.get_icon(), Gtk.IconSize.DIALOG);
+					image_found = true;
+				} else if (app_icon.contains("/")) { // Try to get icon from file
+					var file = File.new_for_uri(app_icon);
+					if (file.query_exists()) {
+						var icon = new FileIcon(file);
+						this.image = new Gtk.Image.from_gicon(icon, Gtk.IconSize.DIALOG);
+						image_found = true;
+					}
+				}
+			}
+
+			// Lastly, for compatibility, check if we have icon_data if no other image was found
+			if (!image_found && (variant = hints.lookup("icon_data")) != null) {
+				var pixbuf = decode_image(variant);
+				if (pixbuf != null) {
+					image = new Gtk.Image.from_pixbuf(pixbuf);
+					image_found = true;
+				}
+			}
```

Comment:
Aren't we already doing this on L132?

Reply:
L132 is for "image-data" and "image_data". Per the spec, "icon_data" is different and has a lower priority. The action is the same, but I'm not sure whether it's worth breaking it into a function, since it's only used in those two places.

Reply:
Oh right, brain misread it as image_data not icon_data, do ignore.